src/Interpreters/InterpreterCreateQuery.cpp

```diff
@@ -1433,6 +1433,25 @@ bool InterpreterCreateQuery::doCreateTable(ASTCreateQuery & create,
 BlockIO InterpreterCreateQuery::doCreateOrReplaceTable(ASTCreateQuery & create,
                                                        const InterpreterCreateQuery::TableProperties & properties)
 {
+    if (create.temporary)
+    {
+        /// Since temporary table does not support RENAME query, so we just simply
+        /// drop old table and create new table.
+        auto ast_drop = std::make_shared<ASTDropQuery>();
```

Comment:
It makes it just a syntax sugar and does not provide atomicity. However, atomicity is not important for temporary tables because no concurrent queries in the same session are possible. It still makes sense to avoid dropping the original tables if the new CREATE fails for some reason, but maybe we can ignore this case (I'm not sure). See also https://github.com/ClickHouse/ClickHouse/issues/31497 (especially https://github.com/ClickHouse/ClickHouse/issues/31497#issuecomment-1416260756)

Reply:
So if we want to make this atomic, we need to first make Database Memory atomic.